### PR TITLE
feat: Refactor validation issues

### PR DESCRIPTION
**Pull request by martinalbert**
## Description

Fixes propagation and formatting of validation issues from map-validator and example-validator.

## Motivation and Context

Multiple duplicated issues at one place and not intuitive messages.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTION_GUIDE** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/interpreter/utils.ts [hunk 1]
```diff
@@ -48,6 +44,73 @@ export function composeVersion(version: VersionStructure): string {
...
+    case 'ListStructure':
+      return `[${formatStructure(structure.value)}]`;
+    case 'NonNullStructure':
+      return `NonNull ${formatStructure(structure.value)}`;
+    case 'ObjectStructure':
+      return `{ ${Object.entries(structure.fields ?? [])
+        .map(([key, type]) => `${key}: ${formatStructure(type)}`)
+        .join(', ')} }`;
+    case 'PrimitiveStructure':
+      return structure.type;
```
Comment (Jakub-Vacek) [inline src/interpreter/utils.ts:72]: maybe formatPrimitive?

Comment (martinalbert) [reply src/interpreter/utils.ts:72]: That is actually string in all cases, its type is `'string' | 'number' | 'boolean'`. Now that you mention it, I always add quotes at `formatIssue` function, so I'll remove `formatPrimitive`, to not cause confusion.

Comment (martinalbert) [reply src/interpreter/utils.ts:72]: Ok, now I understand why I put it there 😄, makes more sense to add quotes to strings when formatting objects. I'll keep `formatPrimitive`, but remove quotes from `formatIssue`.

File: src/interpreter/map-validator.ts [hunk 2]
```diff
@@ -375,8 +380,10 @@ export class MapValidator implements MapAstVisitor {
       // Init new variables if object used in dot notation wasn't defined before
       if (assignment.key.length > 1) {
         const keys = [];
-        for (const key of assignment.key) {
-          keys.push(key);
+        for (let i = 0; i < assignment.key.length - 1; i++) {
```
Comment (Jakub-Vacek) [inline src/interpreter/map-validator.ts:383]: Why did you change this? I

Comment (martinalbert) [reply src/interpreter/map-validator.ts:383]: In old approach, it initializes even whole variable, which is then overwritten because it's assignment anyway.

For example for `some.variable.key = "string"` I wanted to initialize only `some`, `some.variable` and let assignment node handle storing `some.variable.key`.

File: src/interpreter/map-validator.ts [hunk 3]
```diff
@@ -502,11 +480,12 @@ export class MapValidator implements MapAstVisitor {
       const variableName = getVariableName(jessieNode);
       const variable = this.variables[variableName];
 
-      if (variable !== undefined) {
+      if (variable !== undefined && !isInlineCallNode(variable)) {
```
Comment (TheEdward162) [inline src/interpreter/map-validator.ts:483]: What does this change? I don't understand the implications.

Comment (martinalbert) [reply src/interpreter/map-validator.ts:483]: This will prevent propagation of input errors in `inlineCallNode`. I thought about this and it doesn't solve propagation of errors in other three literal nodes, so I'll refactor it and let you know.

Comment (martinalbert) [reply src/interpreter/map-validator.ts:483]: Looks like I won't be able to return that helpful errors without propagation of them, so I'll rather wait for type analysis.

For example in this map:
```
map Test {
  f1 = 1
  f2 = 2

  res = {
    f1: f1,
    f2: f2,
  }

  map result res
}
```

with profile:
```
usecase Test {
  result {
    f1! string!
    f2! boolean!
  }
}
```

I can return all these propagated errors:
```
5:18 PrimitiveLiteral - Wrong Structure: expected string, but got 1
8:19 JessieExpression - Wrong Variable Structure: variable f1 expected string!, but got 1
6:18 PrimitiveLiteral - Wrong Structure: expected boolean, but got 2
8:19 JessieExpression - Wrong Variable Structure: variable f2 expected boolean!, but got 2
13:24 JessieExpression - Wrong Variable Structure: variable res expected {f1: string!, f2: boolean!}!, but got {
    f1: f1,
    f2: f2,
  }
```

or keep only last one and let user understand that `f1` and `f2` are not valid types.

Comment (martinalbert) [reply src/interpreter/map-validator.ts:483]: I would expect that error/warning will show only at the location of map result, but it might be more helpful to have more errors until we don't have type analysis. I'm not sure. @TheEdward162 @Jakub-Vacek what do you think?

Comment (Jakub-Vacek) [reply src/interpreter/map-validator.ts:483]: More errors => higher chance that user will get what is wrong 🤷

**Review by TheEdward162 — APPROVED** [review]
